Create a top level README (#61)
istio/operator: README.md

# Istio Operator

> 

Reviewer: you need a `##` here.  not sure what the `##` should be - possibly `## Introduction`

the `#Istio Operator` won't be displayed I think, but instead will be used as the title of the document.

Istio operator is not yet ready for users to try out (it will be shortly, stay tuned on discuss.istio.io). 
However, the code is stable enough for developers to evaluate and experiment with. Contributions are very welcome - 

> see [open issues](), create a feature proposal, or come to the weekly Environments Working Group meeting if you have ideas. 

Reviewer 2: the open issues link is broken

Reviewer: Can you please integrate my PR (or merge my PR and rebase) here https://github.com/istio/operator/pull/49

I believe this intro still doesn't do what its supposed to do - encourage contribution in an y way shape or form.

Author: Hi @sdake , can you elaborate on why this doesn't encourage contribution (the sentence starts with "Contributions are very welcome" and provides means of doing it, so I'm surprised exactly why you have this reaction). 
Which parts of your PR would you like me to integrate? I believe we agreed to remove the MVP section.

Reviewer: your right my reaction was too strong.  I'd recommend the following improvements:

Put the welcoming text up front.

Please add the following links as you see fit: 
- https://github.com/istio/operator/blob/master/BUGS-AND-FEATURE-REQUESTS.md
- https://github.com/istio/operator/blob/master/CONTRIBUTING.md
- https://github.com/istio/community/blob/master/WORKING-GROUPS.md

> see [open issues](), create a feature proposal, or come to the weekly Environments Working Group meeting if you have ideas. 

Author: Reworded.

> ### Background

Reviewer: use 2 hash not 3.  Might consider a retitle of this section.

Reviewer: WIth the lean and mean approach, which I recommend - the following questions should be answered in a README.md for an open source project

- a call to action inviting participation
- answering why the repo exists
- answering what the repo does
- answering (if desired) how the repo works - this is more suitable for a DESIGN.md rather than the main README.md.

Author: Rigs has created https://github.com/istio/istio/issues/15299 to track this - let's move ahead with this PR since it's time sensitive and defer that work.

The current [helm installation parameters](https://istio.io/docs/reference/config/installation-options/) have been reorganized into two groups:

> 1.  A new [platform level installation API](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto), dealing with k8s settings like resources, auto scaling, pod disruption budgets etc. 

Reviewer: proper md formatting would be 4 spaces, followed by `1. A new`

> 1.  A configuration API, dealing with Istio control plane configuration settings. This API currently uses the [helm installation parameters](https://istio.io/docs/reference/config/installation-options/) for backwards compatibility, but will be reorganized and slimmed down in the future.

Reviewer: dealing with/managing

No need for "but will be reorganized and slimmed down in the future" = who knows what the future holds? :)

> Some parameters will therefore temporarily be in both APIs - for example, setting k8s resources can be done through either API above. However, it's recommended to use the first API, since this has a more consistent structure and it is the one that will remain going forward.  

Reviewer: please remove forward looking statements.
A better way to state this would be "The Istio community recommends usage of the first API as this API will naturally follow the graduation process for APIs while the configuration API is planned for deprecation.

> Typical helm values parameter sets are currently provided as [profiles](https://istio.io/docs/setup/kubernetes/additional-setup/config-profiles/), which act as a starting point for an Istio install, and can currently be customized by editing the values files or passing parameters when calling helm.  

Reviewer: Does typical here mean Pre-existing?

Reviewer: we don't wish to recommend editing the values file at all.  Instead, the values file can be overridden.

> Similarly, the operator API uses the same profiles, which can be selected as a starting point for the installation.   

Reviewer 2: should be "same profiles concept"? The operator API uses profiles concept, but the profile being used is actually different

Author: It is actually the same profiles, just expressed differently. I've updated the wording, hopefully it's clearer now.

For comparison, here is an installation of the SDS profile using helm:

```bash
helm template install/kubernetes/helm/istio --name istio --namespace istio-system \
    --values install/kubernetes/helm/istio/values-istio-sds-auth.yaml | kubectl apply -f -
```

The same installation would be expressed as a CustomResource (CR) through the new API as:

```yaml
# sds-install-cr.yaml

apiVersion: install.istio.io/v1alpha1
kind: IstioControlPlane
metadata:
  name: istio-operator-config
  namespace: istio-system
spec:
  profile: sds
```

The profile can be installed into the cluster by installing the operator into a pod into the cluster and applying the above CR, or applying it via the CLI mode. When using CLI mode, it's not necessary to specify the entire CR, only the spec portion i.e.

```yaml
# sds-install.yaml

profile: sds
```

> In the absence of a profile, the default demo profile is installed. All profiles are available compiled in, or the profile can point to a local file path.

Reviewer 3: ```suggestion
If you don't specify a configuration profile, Istio is installed using the default `demo` configuration profile. 
```
The following sentence is very confusing, please rewrite: "All profiles are available compiled in, or the profile can point to a local file path."

## Developer quick start

> #### Installation 

Reviewer: use 3 hash not 4

Author: Let's park this for now then. We can tackle these issues before 1.3 release.

> The latest, not fully reviewed code is in github.com/ostromart/istio-installer. PRs into the official repo at istio.io/operator are lagging by a few days. 

Reviewer: if this is still true, can you actively speed up the process of cutting over to one repo?  recommend just deleting this text, as it sets a model where there is more than one source of truth.

```bash
git clone https://istio.io/operator.git

> cd istio-installer

Reviewer 2: cd operator

go build -o <your bin path> ./cmd/iop.go
```

#### Flags

> The iop command supports the following flags:

Reviewer 3: ```suggestion
The `iop` command supports the following flags:
```

> +   logtostderr: logs to console (by default logs go to ./iop.go)

Reviewer 3: ```suggestion
+   `logtostderr`: logs to console (by default logs go to ./iop.go)
```

> +   dry-run (default is true for now): console output only, nothing applied to cluster or written to files

Reviewer 3: ```suggestion
+   `dry-run` (default is true for now): console output only, nothing applied to cluster or written to files
```

> +   verbose: display entire manifest contents and other debug info

Reviewer 3: ```suggestion
+   `verbose`: display entire manifest contents and other debug info
```

#### Basic default manifest

```bash
iop manifest
```

> This generates a manifest with all compiled in defaults (charts and base profile). You can see these sources in the repo under data/[profiles|charts]. 

Reviewer 3: It is confusing to readers when you use parenthesis since they can be ambiguous. Please be consistent and introduce the commands with text first. In this case, that means that the text in this line should come before the command on line 75.
```suggestion
The following command generates a manifest with all the default charts and base configuration profiles. You can see these sources in the repository under `data/[profiles|charts]`. 
```

Author: Done.

#### Output to dirs

> The above output is concatenated into a single file. To generate a directory hierarchy, with subdirectory levels representing a child dependency, use:

Reviewer 3: ```suggestion
The output of the manifest is concatenated into a single file. To generate a directory hierarchy with subdirectory levels representing a child dependency, use the following command:
```

```bash
iop manifest -o istio_manifests
```

> This will create a dir hierarchy which should be traversed DFS when applying the yamls. Child manifest directories must wait for the parent, but not sibling manifest directories.

Reviewer 3: ```suggestion
Use Depth Fist Search (DFS) to traverse the created directory hierarchy when applying your YAML files. In other words, child manifest directories must wait for the parent, but not sibling manifest directories.
```

#### Just apply it for me

```bash
iop install
```

> This command will generate the manifests and apply them in the correct dependency order, waiting for dependencies to have CRDs available.   

Reviewer 3: ```suggestion
The following command generates the manifests and applies them in the correct dependency order, waiting for the dependencies to have the needed CRDs available.   
```

> Note: right now the actual "kubectl apply" if only displayed but not run by default. Set --dry-run=false to actually apply to cluster. 

Reviewer 3: Please rewrite the note because it is not clear what it is trying to convey.

Author: Done.

> #### What's in the defaults I installed?

Reviewer 3: ```suggestion
### Review the values of the current configuration profile
```

Author: Done.

> To see the values for the profile in use:

Reviewer 3: ```suggestion
The following command shows the values of the current configuration profile:
```

Author: Done.

```bash
iop dump-profile
```

> This also works with different selected profiles, as in the next section.

Reviewer 3: This text is redundant.

```suggestion
```

Author: Done.

> #### Select a profile

Reviewer 3: ```suggestion
### Select a specific configuration profile
```

Author: Done.

The simplest customization is to select a profile different to default e.g. sds. Create the following config file: 

```yaml
# sds-install.yaml

profile: sds
```

> Pass it to iop:

Reviewer 3: Please do not refer to the Istio operator in the text as "iop". Using undefined abbreviations creates barriers for readers.
```suggestion
Use the Istio operator to apply the new configuration profile:
```

Author: Done.

```bash
iop manifest -f sds-install.yaml
```

> This will cause the helm charts to be rendered with data/profiles/sds.yaml. 

Reviewer 3: ```suggestion
After running the command, the Helm charts are rendered using `data/profiles/sds.yaml`. 
```

Author: Done.

#### Install from file path

> The compiled in charts and profiles are used by default, but a file path (and possibly URL going forward) can be specified, e.g. 

Reviewer: delete forward looking statements.

Author: Done.

# Istio Operator

## Introduction

> The Istio operator CLI is now stable enough for developers to evaluate and experiment with. We welcome your

Reviewer: now stable enough/suitable

Author: Done.

contributions - pick an [unassigned open issue](https://github.com/istio/istio/issues?q=is%3Aissue+is%3Aopen+label%3Aarea%2Fenvironments%2Foperator+no%3Aassignee),
create a feature proposal, or come to the weekly Environments Working Group meeting if you have ideas.

> -  [Bugs and feature requests](https://github.com/istio/operator/blob/master/BUGS-AND-FEATURE-REQUESTS.md)

Reviewer: these types of links work better if they are threaded throughout the introduction text.  I just meant to add the links in some way threaded throughout the text - not a list of links at the end of the section.

Author: Done.

-  [Contributing guidelines](https://github.com/istio/operator/blob/master/CONTRIBUTING.md)
-  [Working groups](https://github.com/istio/community/blob/master/WORKING-GROUPS.md)

### Background

> We reorganized the current [helm installation parameters](https://istio.io/docs/reference/config/installation-options/) into two groups:

Reviewer: We don't use we in docs :).  Instead could I recommend We/This repository/

Author: It's one of the ways to avoid passive voice and be more friendly sounding, but I don't have strong feelings about it.

> -  The new [platform level installation API](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto), for managing

Reviewer: md format is 1 space after `-`., not two.  Two spaces for whatever reason confuses many parsers  You might consider running this README.md by a linter.  A 1 time lint could easily be done by copying it over a file in the docs repo and running the linter there locally.

Author: I installed and ran mdl (which is what istio.io uses) and it now runs clean apart from line length which I'm not going to deal with until I can get a good MD editor.
btw. I changed the number of spaces although that is not flagged by the linter.

> k8s settings like resources, auto scaling, pod disruption budgets etc. 

Reviewer: what is etc here?  Is this basically everything?  IMO sentences should never end in etc as it causes too much cognitive overhead for the reader.  Keep in mind they are reading this for the very first time, the goal is to get them not to think too hard - but to accept the work as is and figure out how to enhance it.  If the README.md is too difficult to parse, they may just drive right by :(

Author: Done.

and will naturally follow the graduation process for APIs while the same parameters in the configuration API are planned
for deprecation.

We currently provide pre-configured helm values sets for different scenarios as configuration 
[profiles](https://istio.io/docs/setup/kubernetes/additional-setup/config-profiles/), which act as a starting point for
an Istio install and can be customized by creating customization overlay files or passing parameters when
calling helm. Similarly, the operator API uses the same profiles (expressed internally through the new API), which can be selected
as a starting point for the installation. For comparison, the following example shows the command needed to install
Istio using the SDS configuration profile using Helm:

```bash
helm template install/kubernetes/helm/istio --name istio --namespace istio-system \
    --values install/kubernetes/helm/istio/values-istio-sds-auth.yaml | kubectl apply -f -
```

In the new API, the same profile would be selected through a CustomResource (CR):

```yaml
# sds-install-cr.yaml

apiVersion: install.istio.io/v1alpha1
kind: IstioControlPlane
metadata:
  name: istio-operator-config

>   namespace: istio-system

Reviewer: i'm confused by this aspect of the API and this is probably a discussion for another thread, however, I'm completely unclear how this permits installation of different components of Istio in different namespaces?

Is the idea the control plane ends up in istio-system?  Some parts of the control plane we want to shave off (citadel for one, and possibly galley) into IIRC istio-base or istio-control.

Author: There's an example further down how to specialize namespaces by feature and/or component. But we can take another pass at this doc at a high level to address such gaps.

spec:
  profile: sds
```

> CRs are used when running the operator as a controller in a pod in the cluster. When using operator CLI mode and passing the 

Reviewer: this could probably be simplified such as "CRs are consumed by the operator when it has been started in Kubernetes via a manifest.

controller is some pretty serious lingo for an end-user..

Author: It may not be clear that these docs are for devs, not users. The operator is not ready for users, only devs and one of the missing parts is proper user documentation.

> configuration as a file (see [Select a profile](#Select_a_profile)), only the spec portion is required.

Reviewer: something wrong with this sentence.  please try to reword.  Anytime a `(` is needed in a sentence, the sentence can be reworked into two sentences.  Atleast use a `;` instead.

```yaml
# sds-install.yaml

profile: sds
```

> If you don't specify a configuration profile, Istio is installed using the `default` configuration profile. All 

Reviewer: I think you should lead with this text, rather than leading with the complexity of installing an SDS profile.

Author: We can discuss this, but if this is omitted, there's nothing to pass to the CLI which is weird in this case. The first example in the walkthrough does cover that case, but it may be worth rethinking the layout.

profiles listed in istio.io are available compiled in, or `profile:` can point to a local file path to reference a custom
profile base to use as a starting point for customization. See the [API reference](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto)
for details.

> ## Developer quick start

Reviewer: although I disagree with the choice, I think project wide, the docs WG has made a decision that developer material belongs in the consolidated wiki.  @rcaballeromx is this accurate?

The quick start describes how to install and use the operator `iop` CLI command. 

#### Installation 

```bash
git clone https://istio.io/operator.git
cd operator

> go build -o <your bin path> ./cmd/iop.go

Reviewer: we have found with cut and paste, alot of cut and paste ends up with erroneous problems because of substitutions.  I'd recommend using "$GOPATH/bin" here.

Author: Sg, thanks, done.

```

#### Flags

The `iop` command supports the following flags:

> +   `logtostderr`: log to console (by default logs go to ./iop.go).

Reviewer: bullets can use `+`, `-`, or `*` - but they should be consistent across the docs.  I've been working hard to steer people towards `-` in the documentation repository.

Author: Changed to -.

+   `dry-run`: console output only, nothing applied to cluster or written to files (default is true for now).
+   `verbose`: display entire manifest contents and other debug info (default is false).

#### Basic default manifest

The following command generates a manifest with the compiled in default profile and charts:

```bash
iop manifest
```

> You can see these sources for the compiled in profiles in the repo under data/profiles, while the compiled in helm

Reviewer: data/profiles and data/charts should probably be in backticks

Author: Done.

charts are under data/charts.

#### Output to dirs

The output of the manifest is concatenated into a single file. To generate a directory hierarchy with subdirectory
levels representing a child dependency, use the following command:

```bash
iop manifest -o istio_manifests
```

> Use depth first search to traverse the created directory hierarchy when applying your YAML files. Child manifest

Reviewer: this block could be simplified...

Reviewer 3: ```suggestion
When you apply your YAML files, use depth first search to traverse the created directory hierarchy. This is needed because child manifest
```

Author: The suggestion has a slightly different meaning - I've updated the sentence to address what I think you're trying to address here.

directories must wait for the parent, but not sibling manifest directories.

#### Just apply it for me

The following command generates the manifests and applies them in the correct dependency order, waiting for the
dependencies to have the needed CRDs available: 

```bash
iop install
```

> Note: right now the configuration that would be applied is only displayed, since `dry-run` is true by default. Set

Reviewer: delete right now and capitalize `T`he.

Reviewer: reorder the sentence such that the `dry-run=true` is started up front.  E.g. : `The default dry-run=true will force the display of the manifest only.  Set dry-run=false to apply the manifest to Istio.`

Reviewer 3: ```suggestion
The default `dry-run=true` displays the manifest only. Set `dry-run=false` to apply the Istio configuration to Kubernetes.
```

`--dry-run=false` to actually apply the generated configuration to the cluster.

#### Review the values of the current configuration profile

The following command shows the values of the current configuration profile:

```bash
iop dump-profile
```

#### Select a specific configuration profile

The simplest customization is to select a profile different to `default` e.g. `sds`. Create the following config file: 

```yaml
# sds-install.yaml

profile: sds
```

Use the Istio operator `iop` binary to apply the new configuration profile:

```bash
iop manifest -f sds-install.yaml
```

After running the command, the Helm charts are rendered using `data/profiles/sds.yaml`. 

#### Install from file path

> The compiled in charts and profiles are used by default, but a file path can be specified, e.g.

Reviewer: `, e.g. ` replace with `:`

```yaml
profile: file:///usr/home/bob/go/src/github.com/ostromart/istio-installer/data/profiles/default.yaml
customPackagePath: file:///usr/home/bob/go/src/github.com/ostromart/istio-installer/data/charts/
```

> These can be mixed and matched e.g. use a compiled in profile with local filesystem charts. 

Reviewer: What does `These` refer to?  e.g. ideally shouldn't be in the middle of a sentence.  If this is accurate, you might go with something like "Multiple file paths may be specified to enable the use of both the compiled in profile in addition to local filesystem charts.`

Author: Done.

#### New API customization

The [new platform level installation API](https://github.com/istio/operator/blob/95e89c4fb838b0b374f70d7c5814329e25a64819/pkg/apis/istio/v1alpha1/istioinstaller_types.proto#L25) 

> defines install time parameters like feature/component enablement and namespace, and k8s settings like resources, HPA spec etc. in a structured way.   

Reviewer: /feature/component -> feature and component

Author: Done.

The simplest customization is to turn features and components on and off. For example, to turn off all policy:

```yaml
profile: sds
policy:
  enabled: false
```

> Note that unlike helm, all configurations are validated against a schema, so the operator detects syntax errors. Another

Reviewer: capitalize Helm.

Reviewer: I'd reword this sentence as follows "The operator validates the configuration and automatically detects syntax errors.  Helm lacks this capability.  If you are using Helm values that are incompatible, the schema validation used in the operator may likely reject input that is valid for Helm."

Author: Capitalized throughout. 
Re: the rewording, that's a great suggestion. Thanks.

customization is to define custom namespaces for features:

```yaml
profile: sds

trafficManagement:
  namespace: istio-control-custom 
```

> The traffic management feature comprises Pilot, AutoInjection and Proxy components. Each of these components has k8s

Reviewer: Throughout document, either change k8s->K8s or Kubernetes.  K8s is the correct trademark way to communicate Kubernetes, although spelling it out looks more prfessional IMO.

Author: Done. I agree Kubernetes looks better but it's used so much in the doc, I went with K8s.

> settings, and these can be overridden from the defaults using official k8s APIs (rather than Istio defined schemas):

Reviewer: this point is hard to understand.  Perhaps too much detail?

Author: I definitely want to make this point somewhere; I honestly couldn't think of a way to simplify this. Let's stick with this for now, and maybe we can improve on the docs as a whole when we take a pass on them before the release.

+   [readiness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/)
+   [replica count](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/)
+   [HoriizontalPodAutoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
+   [PodDisruptionBudget](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#how-disruption-budgets-work)
+   [pod annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/)
+   [service annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/)
+   [ImagePullPolicy](https://kubernetes.io/docs/concepts/containers/images/)
+   [priority calss name](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#priorityclass)
+   [node selector](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector)
+   [affinity and anti-affinity](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity)

> All of these k8s settings use the k8s API definitions, so k8s documentation can be used for reference. All k8s overlay

Reviewer: link the k8s documentationn to docs/concepts page?

Author: Done.

values are also validated in the operator.

#### Customizing the old values.yaml API 

The new platform install API above deals with k8s level settings. The remaining values.yaml parameters deal with Istio

> control plane operation rather than platform installation and for the time being the operator just passes these through

Reviewer: this sentence would benefit from periods :)

Author: Done.

The quick start describes how to install and use the operator `iop` CLI command.

#### Installation

```bash
git clone https://istio.io/operator.git
cd operator
go build -o <your bin path> ./cmd/iop.go
```

#### Flags

The `iop` command supports the following flags:

+ `logtostderr`: log to console (by default logs go to ./iop.go).
+ `dry-run`: console output only, nothing applied to cluster or written to files (default is true for now).
+ `verbose`: display entire manifest contents and other debug info (default is false).

#### Basic default manifest

The following command generates a manifest with the compiled in default profile and charts:

```bash
iop manifest
```

You can see these sources for the compiled in profiles in the repo under `data/profiles`, while the compiled in Helm
charts are under `data/charts`.

#### Output to dirs

The output of the manifest is concatenated into a single file. To generate a directory hierarchy with subdirectory
levels representing a child dependency, use the following command:

```bash
iop manifest -o istio_manifests
```

Use depth first search to traverse the created directory hierarchy when applying your YAML files. Child manifest

> directories must wait for the parent, but not sibling manifest directories.

Reviewer 3: ```suggestion
directories must wait for their parent directory, but not for their sibling manifest directories.
```

dependencies to have the needed CRDs available:

```bash
iop install
```

Note: right now the configuration that would be applied is only displayed, since `dry-run` is true by default. Set
`--dry-run=false` to actually apply the generated configuration to the cluster.

#### Review the values of the current configuration profile

The following command shows the values of the current configuration profile:

```bash
iop dump-profile
```

#### Select a specific configuration profile

The simplest customization is to select a profile different to `default` e.g. `sds`. Create the following config file:

```yaml
# sds-install.yaml

profile: sds
```

Use the Istio operator `iop` binary to apply the new configuration profile:

```bash
iop manifest -f sds-install.yaml
```

After running the command, the Helm charts are rendered using `data/profiles/sds.yaml`.

#### Install from file path

The compiled in charts and profiles are used by default, but a file path can be specified, e.g.

```yaml
profile: file:///usr/home/bob/go/src/github.com/ostromart/istio-installer/data/profiles/default.yaml
customPackagePath: file:///usr/home/bob/go/src/github.com/ostromart/istio-installer/data/charts/
```

> These can be mixed and matched e.g. use a compiled in profile with local filesystem charts.

Author: Done, with minor change from suggestion.

Helm. Another customization is to define custom namespaces for features:

```yaml
profile: sds

trafficManagement:
  namespace: istio-control-custom
```

The traffic management feature comprises Pilot, AutoInjection and Proxy components. Each of these components has K8s
settings, and these can be overridden from the defaults using official K8s APIs (rather than Istio defined schemas):

```yaml
trafficManagement:
  components:
    pilot:
      k8s:
        resources:
          requests:
            cpu: 1000m # override from default 500m
            memory: 4096Mi # ... default 2048Mi
        hpaSpec:
          maxReplicas: 10 # ... default 5
          minReplicas: 2  # ... default 1
```

The K8s settings are defined in detail in the
[operator API](https://github.com/istio/operator/blob/95e89c4fb838b0b374f70d7c5814329e25a64819/pkg/apis/istio/v1alpha1/istioinstaller_types.proto#L394).
The settings are the same for all components, so a user can configure pilot K8s settings in exactly the same, consistent
way as galley settings. Supported K8s settings currently include:

> + [resources](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#resource-requests-and-limits-of-pod-and-container)

Author: Done.
